File: src/data/courses/dsa/stacks/stack-basic-applications/evaluate-postfix-prefix/code/java/EvaluatePostfixPrefix.java
// ============================================================
// Evaluate Reverse Polish Notation
// ============================================================

import java.util.*;

class Solution {
    public int evalRPN(String[] tokens) {
        Deque<Long> stack = new ArrayDeque<>();
        for (String t : tokens) {
            if ("+-*/".contains(t) && t.length() == 1) {
                long b = stack.pop(), a = stack.pop();
                switch (t) {
                    case "+": stack.push(a + b); break;
                    case "-": stack.push(a - b); break;
                    case "*": stack.push(a * b); break;
                    case "/": stack.push(a / b); break;
                }
            } else {
                stack.push(Long.parseLong(t));
            }
        }
        return stack.pop().intValue();
    }
}

File: src/data/courses/dsa/stacks/stack-basic-applications/evaluate-postfix-prefix/code/python/evaluate_postfix_prefix.py
# ============================================================
# Evaluate Reverse Polish Notation
# ============================================================

from typing import List

class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        stack = []
        ops = {'+', '-', '*', '/'}

        for t in tokens:
            if t in ops:
                b, a = stack.pop(), stack.pop()  # b first, a second
                if t == '+': stack.append(a + b)
                elif t == '-': stack.append(a - b)
                elif t == '*': stack.append(a * b)
                else: stack.append(int(a / b))  # Truncate toward zero
            else:
                stack.append(int(t))

        return stack[0]
